functorch/_src/aot_autograd.py

```diff
@@ -167,86 +181,336 @@ def posthook(grad_input, grad_output):
+                    if out_storage_ref in inp_storage_refs:
+                        maybe_aliased_inp_idx = inp_storage_refs[out_storage_ref]
+                    else:
+                        maybe_aliased_inp_idx = None
+                else:
+                    maybe_aliased_inp_idx = None
+
+                if maybe_aliased_inp_idx is None:
+                    # Only return outputs that are not aliases of inputs.
+                    non_aliased_outs.append(o)
+
+                # Also, track the metadata to properly regenerate these outputs later.
+                # We store (inp_idx, fake_tensor_representing_output)
+                # That way later we can regenerate the output with inputs[inp_idx].as_strided(fake_tensor_output_metadata)
+                if maybe_aliased_inp_idx is None:
+                    # Note that if we have something like:
+                    # def f(x):
+                    #     x.add_(1)
+                    #     return x
+                    # Our compiled fw will return an "x_updated", but it is *not* ok to return that to the user.
+                    # We need to manually do x.copy_(x_updated), and return the original x to the user.
+                    # Why? for example, the metadata between x and x_updated might be different (e.g. _is_leaf())
+                    collect_aliased_output_info.append(None)
+                elif id(o) in inp_tensor_ids:
+                    # Special case: the output *was* an input. Don't regenerate a view for it, just return the input directly
+                    # This is signaled by the fact that the second tuple arg is None
+                    collect_aliased_output_info.append((maybe_aliased_inp_idx, None))
+                else:
+                    collect_aliased_output_info.append((maybe_aliased_inp_idx, o))
+            return non_aliased_outs
+
+        non_aliased_outs = filter_and_record_aliased_outs(outs)
+
+        pytree.tree_map(collect_grad_info, non_aliased_outs)
+
+        # Calling convention: the output is (mutated_input_values, original_outs)
+        # We return all mutated inputs + outputs here, **except** for any mutated inputs or outputs
+        # that alias original inputs.
+        # See Note [Input mutation handling in aot autograd]
+        mutated_inps_and_outs = inputs_with_mutated_data + list(non_aliased_outs)
+
+        # Our autograd.Function.forward returns both mutated inputs and outputs,
+        # so we need grad info on all of them.
+        assert len(collect_requires_grad_out_info) == len(mutated_inps_and_outs)
+
+        metadata = ViewAndMutationMeta(
+            mutated_input_info=collect_mutated_input_info,
+            metadata_mutation_input_info=collect_metadata_mutation_input_info,
+            requires_grad_out_info=collect_requires_grad_out_info,
+            aliased_output_info=collect_aliased_output_info,
+        )
+        return metadata, pytree.tree_map(from_fun, mutated_inps_and_outs)
```

Comment:
NB: calling convention change was avoided by computing metadata in the preflight. This seems fine.